Weighted vehicle spawn pool for airbases

wersal454 opened the pull request:
## What type of PR is this?
1. [ ] Bug
2. [ ] Change
3. [ ] Enhancement
4. [x] Miscellaneous

### What have you changed and why?
Information:

that was a lot of effort, but now you can adjust spawn chances for each vehicle category

### Please specify which Issue this PR Resolves (If Applicable).
"This PR closes #XXXX!"

### Please verify the following.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [x] No
2. [ ] Yes (suggest weights for each category)

### How can the changes be tested?
Steps:

********************************************************
Notes:
A lot of effort

wersal454 committed:
Update fn_createAIAirplane.sqf

wersal454 committed:
weighted vehicle spawn pool

wersal454 commented:
UAV exist in CAS category because, they can drop bombs alright, though deleting them from AA is a good idea, I never really tested how they perform as interceptors

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
Merge branch 'unstable' into patch-52

Silence-112 reviewed (commented):
It seems to work as expected, I do think that the weights should be changed a bit:

Hangar:
PlanesAA/CAS: 1
UAV: 0.3

Runway:
LargeAA/CAS, Transport: 1
PlanesAA/CAS: 0.7
Gunship: 0.4

Also I think it'd be a good idea to introduce a new parameter for UAV spawns at airbases, should be easy enough to skip adding them if it's disabled

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 commented:
I thought about adding weights to ground vehicles like in #3440

wersal454 committed:
adjustments

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
UAV spawn chance parametr

wersal454 committed:
partially ported #3440

wersal454 commented:
done

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
Update Stringtable.xml

Silence-112 commented on A3A/addons/core/Params.hpp:
```diff
@@ -785,9 +785,17 @@ class Params
         tooltip = $STR_params_rebelGarrisonLimit_desc;
         values[] = {-1, 16, 24, 32};
-        texts[] = {"No limit", "16", "24", "32"};
+        texts[] = {"∞", "16", "24", "32"};
```
I think "No limit" is probably better here, wouldn't surprise me if this doesn't get displayed or interpreted properly

wersal454 replied:
I don't think unlocalized "No limit" is a good idea

Silence-112 replied:
Change it in another pull. It's not really a relevant change in this PRs scope

wersal454 replied:
![image](https://github.com/user-attachments/assets/242185a0-74fd-4f23-9e55-64c5cfc66aef)
it works btw

Silence-112 commented on A3A/addons/scrt/Stringtable.xml:
```diff
@@ -5514,6 +5514,14 @@
                 <French>Nombre de troupes rebelles pouvant se trouver en garnison. Notez que l'option Pas de limite peut affecter les performances car il pourrait y avoir trop de troupes pour que le CPU puisse les gérer.</French>
             </Key>
+            <Key ID="STR_params_UAVSpawnChance">
+                <Original>UAV spawn chance on airfileds</Original>
```
```suggestion
                <Original>UAV spawn chance at airfields</Original>
```

Silence-112 commented on A3A/addons/scrt/Stringtable.xml:
```diff
@@ -5514,6 +5514,14 @@
+                <Russian>Шанс спавна БПЛА на аэрдромах</Russian>
+            </Key>
+            <Key ID="STR_params_UAVSpawnChance_desc">
+                <Original>Adjusts spawn chance of UAVs on enemy airfields, higher the chance, less likely the game will spawn other fixed wing aircrafts</Original>
```
```suggestion
                <Original>Adjusts spawn chance of UAVs at enemy airfields. The higher the chance, the less likely other fixed wing aircrafts will be spawned</Original>
```

Silence-112 commented on A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf:
```diff
@@ -398,18 +440,39 @@ if (!_busy) then
-	(_faction get "vehiclesFuelTrucks") + 
-	(_faction get "vehiclesMedical");
-_countX = 0;
+private _lightvehPool = [];
+private _vehTypes = [
+	"vehiclesLightArmed",
+	"vehiclesLightUnarmed",
+	"vehiclesTrucks",
+	"vehiclesCargoTrucks",
+	"vehiclesAmmoTrucks",
+	"vehiclesRepairTrucks",
+	"vehiclesFuelTrucks",
+	"vehiclesMedical"
+	];
+private _typeWeight = [
+	7, 
+	4, 
+	2, 
+	2,
+	1, 
+	1, 
+	1, 
+	2
+	];
```
```suggestion
private _groundPool = [];

private _vehTypes = [
	"vehiclesLightArmed",
	"vehiclesLightUnarmed",
	"vehiclesTrucks",
	"vehiclesCargoTrucks",
	"vehiclesAmmoTrucks",
	"vehiclesRepairTrucks",
	"vehiclesFuelTrucks",
	"vehiclesMedical"
];

private _vehTypeWeights = [
	7, 
	4, 
	2, 
	2,
	1, 
	1, 
	1, 
	2
];
```

Silence-112 commented on A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf:
```diff
@@ -398,18 +440,39 @@ if (!_busy) then
+	2
+	];
+
+{
+	private _vehs = _faction get _x;
+	if(_vehs isEqualTo []) then {continue};
+	private _weight = (_typeWeight select _forEachIndex) / count _vehs;
+	{
+		_lightvehPool append [_x, _weight];
+	} forEach _vehs;
+} forEach _vehTypes;
```
```suggestion
{
	private _vehs = _faction get _x;
	if (_vehs isEqualTo []) then {continue};
	private _weight = (_vehTypeWeights select _forEachIndex) / count _vehs;
	{
		_groundPool append [_x, _weight];
	} forEach _vehs;
} forEach _vehTypes;
```

Silence-112 commented on A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf:
```diff
@@ -398,18 +440,39 @@ if (!_busy) then
 
 while {_countX < _nVeh && {_countX < 3}} do {
-	private _typeVehX = selectRandom _vehTypesLight;
+	private _typeVehX = selectRandomWeighted _lightvehPool;
```
```suggestion
	private _typeVehX = selectRandomWeighted _groundPool;
```

wersal454 committed:
Update A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf

Co-authored-by: Silence <78276788+SilenceIsFatto@users.noreply.github.com>

wersal454 committed:
Update A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf

Co-authored-by: Silence <78276788+SilenceIsFatto@users.noreply.github.com>

wersal454 committed:
Update A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf

Co-authored-by: Silence <78276788+SilenceIsFatto@users.noreply.github.com>

wersal454 committed:
Update Stringtable.xml

wersal454 committed:
Update Stringtable.xml

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
Update Params.hpp

wersal454 committed:
fixes

wersal454 committed:
Merge branch 'patch-52' of https://github.com/wersal454/A3-Antistasi into patch-52

Silence-112 commented on A3A/addons/core/functions/CREATE/fn_createAIAirplane.sqf:
```diff
@@ -299,41 +299,87 @@ if (!_busy) then {
+						_vehPool pushBack (A3A_UAVSpawnChance);
+					} else {
+						_vehPool pushBack (A3A_UAVSpawnChance - 0.1);
+					};
```
```suggestion
					_vehPool pushBack ((A3A_UAVSpawnChance - 0.1) max 0);
```

wersal454 committed:
Update fn_createAIAirplane.sqf

wersal454 committed:
weighted milbase veh spawn

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
Merge branch 'unstable' into patch-52

wersal454 committed:
fixes + changed AA and heavy patrol vehicle spawn logic

AA and heavy patrol vehicles now spawn last to avoid explosive collison situations
+ heavy patrol vehicle is actually doing the patrol thing
+ fixed (I hope) issue where AA and heavy patrol vehicles weren't getting deleted when player leaves the area
+ statics won't be locked anymore
